hazelcast-client/src/main/java/com/hazelcast/client/proxy/ClientFlakeIdGeneratorProxy.java

```diff
@@ -0,0 +1,233 @@
+ * limitations under the License.
+ */
+
+package com.hazelcast.client.proxy;
+
+import com.hazelcast.client.config.FlakeIdGeneratorConfig;
+import com.hazelcast.client.impl.ClientMessageDecoder;
+import com.hazelcast.client.impl.protocol.ClientMessage;
+import com.hazelcast.client.impl.protocol.codec.FlakeIdGeneratorNewIdBatchCodec;
+import com.hazelcast.client.impl.protocol.codec.FlakeIdGeneratorNewIdBatchCodec.ResponseParameters;
+import com.hazelcast.client.spi.ClientContext;
+import com.hazelcast.client.spi.ClientProxy;
+import com.hazelcast.client.spi.impl.ClientInvocation;
+import com.hazelcast.client.spi.impl.ClientInvocationFuture;
+import com.hazelcast.client.util.ClientDelegatingFuture;
+import com.hazelcast.concurrent.flakeidgen.FlakeIdNodeIdOverflowException;
+import com.hazelcast.core.FlakeIdGenerator;
+import com.hazelcast.core.HazelcastException;
+import com.hazelcast.core.IFunction;
+import com.hazelcast.core.IdGenerator;
+import com.hazelcast.core.Member;
+import com.hazelcast.internal.util.ThreadLocalRandomProvider;
+import com.hazelcast.nio.Address;
+import com.hazelcast.spi.InternalCompletableFuture;
+import com.hazelcast.util.Clock;
+import com.hazelcast.util.ExceptionUtil;
+
+import java.util.ArrayList;
+import java.util.List;
+import java.util.Set;
+import java.util.concurrent.ConcurrentHashMap;
+import java.util.concurrent.atomic.AtomicIntegerFieldUpdater;
+
+import static java.util.Collections.newSetFromMap;
+
+/**
+ * Proxy implementation of {@link IdGenerator}.
+ */
+public class ClientFlakeIdGeneratorProxy extends ClientProxy implements FlakeIdGenerator {
+
+    private static final int DEFAULT_NUM_ATTEMPTS = 5;
+
+    // this should be final, but is not due to tests
+    private static int numAttempts = DEFAULT_NUM_ATTEMPTS;
+
+    private static final ClientMessageDecoder NEW_ID_BATCH_DECODER = new ClientMessageDecoder() {
+        @Override
+        public IdBatch decodeClientMessage(ClientMessage clientMessage) {
+            ResponseParameters response = FlakeIdGeneratorNewIdBatchCodec.decodeResponse(clientMessage);
+            return new IdBatch(response.base, response.increment, response.batchSize);
+        }
+    };
+
+    /**
+     * Set of member UUIDs of which we know have overflowed NodeIds. These members are never again tried
+     * to get ID from.
+      */
+    private final Set<String> overflowedMembers = newSetFromMap(new ConcurrentHashMap<String, Boolean>(0));
+
+    /**
+     * Current randomly chosen member from which we are getting IDs.
+     */
+    private volatile Member randomMember;
+
+    private AutoBatcher batcher;
+
+    public ClientFlakeIdGeneratorProxy(String serviceName, String objectName, ClientContext context) {
+        super(serviceName, objectName, context);
+    }
+
+    @Override
+    protected void onInitialize() {
+        super.onInitialize();
+
+        FlakeIdGeneratorConfig config = getContext().getClientConfig().findFlakeIdGeneratorConfig(getName());
```

Comment:
It's inspired by other usages `onInitialize`, for example `NearCachedClientMapProxy`, `ClientReplicatedMapProxy` etc.

Reply:
I see. The only difference between the two is that if the constructor throws an exception, the proxy creation [fails immediately](https://github.com/hazelcast/hazelcast/blob/611c7505c8bcb310d00b987d4e3702610a298d1e/hazelcast-client/src/main/java/com/hazelcast/client/spi/ProxyManager.java#L312-L321) while if the `onInitialize` method throws an exception, it [might be retried](https://github.com/hazelcast/hazelcast/blob/611c7505c8bcb310d00b987d4e3702610a298d1e/hazelcast-client/src/main/java/com/hazelcast/client/spi/ProxyManager.java#L337-L361). Since I don't see a reason why the `onInitialize` method should be retried here, I suggest the initialisation be moved to the constructor. This way, we avoid having an additional method in this proxy.

Reply:
That's a good finding! Thanks for digging into this :)